core/src/main/scala/org/apache/spark/sql/sources/SparkOptimizations.scala

```diff
@@ -0,0 +1,410 @@
+/*
+ * Copyright (c) 2016 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+package org.apache.spark.sql.sources
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+import com.gemstone.gemfire.internal.cache.{AbstractRegion, ColocationHelper, PartitionedRegion}
+import com.pivotal.gemfirexd.internal.engine.Misc
+import io.snappydata.QueryHint
+
+import org.apache.spark.sql.SnappySession
+import org.apache.spark.sql.catalyst.analysis.{EliminateSubqueryAliases, UnresolvedAlias, UnresolvedAttribute, UnresolvedFunction, UnresolvedGenerator, UnresolvedStar}
+import org.apache.spark.sql.catalyst.expressions
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, AttributeSet, BinaryComparison, Cast, Expression, PredicateHelper, UnresolvedWindowExpression}
+import org.apache.spark.sql.catalyst.planning.ExtractEquiJoinKeys
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, SubqueryAlias}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.columnar.impl.{BaseColumnFormatRelation, ColumnFormatRelation, IndexColumnFormatRelation}
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+
+
+/**
+  * Replace table with index hint
+  */
+case class ResolveQueryHints(snappySession: SnappySession) extends Rule[LogicalPlan]
```

Comment:
It will be good to have to compare the index performance against the table performance.